src/cascadia/TerminalApp/TabHeaderControl.idl

```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+namespace TerminalApp
+{
+    delegate void HeaderTitleChangedArgs(String title);
+
+    [default_interface] runtimeclass TabHeaderControl : Windows.UI.Xaml.Controls.UserControl
+    {
+        String Title { get; };
+
+        TabHeaderControl();
+        void UpdateHeaderText(String title);
+        void SetZoomIcon(Windows.UI.Xaml.Visibility state);
```

Comment:
Here's a crazy idea:

Could we just pass a `TabBase` to the `TabHeaderControl`, and then bind to the header's `tabBase.Title` directly? That would avoid the need to pass the tab title back and forth.

Plus, then we could use a converter to convert the `TabBase`'s zoom state into a `Visibility` for the zoom icon, rather than manually setting the visibility of the icon when we zoom in/out. Thoughts?

Reply:
I feel like this increases coupling for little good.

Right now, TabHeaderControl is a control that can stand alone and display state information from _a thing_. It supports renaming and displaying state info. It might eventually be a good basis for a PaneTitleBar's Content because of that.

If we make it depend on TabBase, not only are we coupling TabBase and TabHeaderControl, we're introducing a 2-way dependency s.t. a TabHeaderControl holds on to a reference to its TabBase and TabBase owns a TabHeaderControl.

If we keep it separate, I am imagining a world where we can do this (in xaml) ...

```xaml
<local:TabHeaderControl
 Title="{x:Bind ActiveTerminalTitle, Mode=OneWay}"
 ZoomIndicatorVisible="{x:Bind IsPaneZoomed, Mode=OneWay}"
 ...
 />
```

I know it turns the ferret inside-out and makes us hook up a lot of dependent properties, but it feels cleaner than every control reaching around into its parent to ask questions that it could be told.

Reply:
Okay, I've been convinced.

Reply:
It's for this reason that I encouraged Pankaj to propagate an event for a rename request. If we live update the title in the HeaderControl, we are (even momentarily!) giving it _control_ over the title[2]. It should never have that control, because it cannot see from where it is sitting how it plays a part in the bigger picture[1]. It violates the principle that this is a dumb data container.

What do you think?

[1] If the tab doesn't want the title to have changed (for various reasons), it has to listen to the "a property changed event" and _revert it_ if it was the title.

[2] The tab needs to know about what the user-overridden title is _anyway_ so that it wins out over pane titles or app titles; there's no value in having two masters for the title.

Reply:
OOP, I didn't see your convincement when I was writing up the followup message. :grin:

Reply:
_I'm now thoroughly convinced_